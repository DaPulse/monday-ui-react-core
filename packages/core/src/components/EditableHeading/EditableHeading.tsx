import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import React, { forwardRef } from "react";
import Heading from "../Heading/Heading";
import { VibeComponent, VibeComponentProps, withStaticProps } from "../../types";
import styles from "./EditableHeading.module.scss";
import { HeadingType as HeadingTypeEnum, HeadingWeight as HeadingWeightEnum } from "../Heading/HeadingConstants";
import { HeadingType, HeadingWeight } from "../Heading/Heading.types";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import { camelCase } from "lodash-es";
import EditableTypography, { EditableTypographyImplementationProps } from "../EditableTypography/EditableTypography";

export interface EditableHeadingProps extends VibeComponentProps, EditableTypographyImplementationProps {
  /**
   * Sets the Heading type
   * @type {HeadingType}
   * */
  type?: HeadingType;
  /** Sets the Heading weight
   * @type {HeadingWeight}
   */
  weight?: HeadingWeight;
}

const EditableHeading: VibeComponent<EditableHeadingProps, HTMLElement> & {
  types?: typeof HeadingTypeEnum;
  weights?: typeof HeadingWeightEnum;
} = forwardRef(
  (
    { type = "h1", weight = "normal", id, "data-testid": dataTestId, ...editableTypographyProps }: EditableHeadingProps,
    ref
  ) => {
    return (
      <EditableTypography
        ref={ref}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.EDITABLE_HEADING, id)}
        component={Heading}
        typographyClassName={getStyle(styles, camelCase(type + "-" + weight))}
        type={type}
        weight={weight}
        {...editableTypographyProps}
      />
    );
  }
);

export default withStaticProps(EditableHeading, {
  types: HeadingTypeEnum,
  weights: HeadingWeightEnum
});
