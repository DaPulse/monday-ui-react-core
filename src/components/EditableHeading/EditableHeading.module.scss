@import "../../styles/typography";
@import "~monday-ui-style/dist/mixins";

@include create-title-classes;

.editableHeading {
  display: inline-flex;
  width: 100%;
  min-width: 0;

  .input {
    display: inline-block;
    max-width: 100%;
    min-width: 64px;
    padding: 0 var(--spacing-xs);
    outline: none;
    border: 1px solid var(--primary-color);
    border-radius: var(--border-radius-small);

    &:focus,
    &:active {
      outline: none;
    }
  }

  .heading {
    border: 1px solid transparent;
    padding: 0 var(--spacing-xs);
    border-radius: var(--border-radius-small);

    &:hover:not(.disabled) {
      border-color: var(--ui-border-color);
    }
  }
}
