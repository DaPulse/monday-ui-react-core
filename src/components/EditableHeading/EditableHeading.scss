@import "../../styles/themes";
@import "../../styles/typography";
@import "../../styles/global-css-settings";
@import "../../styles/states";

.editable-heading--wrapper {
  width: 100%;
  min-width: 0;
  border: 1px solid transparent;
  border-radius: $border-radius-small;
  @include focus-style();
  &:focus,
  &:active {
    outline: none;
  }

  &.inset-focus {
    @include focus-style-inset();
  }

  .font-weight {
    &-light {
      font-weight: 300;
    }
    &-normal {
      font-weight: 500;
    }
    &-bold {
      font-weight: 700;
    }
  }

  .heading-component {
    margin: 0;
  }
}
