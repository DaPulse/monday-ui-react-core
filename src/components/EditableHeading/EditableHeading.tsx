import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import React, { forwardRef, useEffect, useLayoutEffect, useMemo, useRef, useState } from "react";
import VibeComponentProps from "../../types/VibeComponentProps";
import Heading from "../Heading/Heading";
import { withStaticProps } from "../../types";
import styles from "./EditableHeading.module.scss";
import { HeadingType, HeadingWeight } from "../Heading/HeadingConstants";
import cx from "classnames";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import { camelCase } from "lodash-es";
import HiddenInputPlaceholder from "./HiddenInputPlaceholder";
import { keyCodes } from "../../constants";
import useMergeRefs from "../../hooks/useMergeRefs";
import useClickableProps from "../../hooks/useClickableProps/useClickableProps";

export interface EditableHeadingProps extends VibeComponentProps {
  /** Heading text */
  value: string;
  /** Heading type */
  type?: HeadingType;
  /** Heading weight */
  weight?: HeadingWeight;
  /** onChange callback */
  onChange?: (value: string) => void;
  /** Disable editing mode */
  readOnly?: boolean;
  /** Focus input on mount in edit mode */
  focusOnMount?: boolean;
  /** Select input text on mount in edit mode */
  selectOnMount?: boolean;
  /** ARIA Label */
  ariaLabel?: string;
}

const EditableHeading: React.FC<EditableHeadingProps> & {
  types?: typeof Heading.types;
  weights?: typeof Heading.weights;
} = forwardRef(
  (
    {
      id,
      className,
      "data-testid": dataTestId,
      value,
      type = Heading.types.H1,
      weight = HeadingWeight.NORMAL,
      onChange,
      readOnly = false,
      focusOnMount = true,
      selectOnMount = false,
      ariaLabel = ""
    },
    ref
  ) => {
    const componentRef = useRef(null);
    const mergedRef = useMergeRefs({ refs: [ref, componentRef] });

    const [isEditing, setIsEditing] = useState(false);
    const [inputValue, setInputValue] = useState(value);
    const [inputWidth, setInputWidth] = useState(0);

    const inputRef = useRef(null);

    function toggleEditMode() {
      if (readOnly || isEditing) {
        return;
      }

      setIsEditing(true);
    }

    function handleInputValueChange() {
      setIsEditing(false);
      if (!inputValue || value === inputValue) {
        setInputValue(value);
        return;
      }
      onChange?.(inputValue);
    }

    function handleBlur() {
      handleInputValueChange();
    }

    function handleKeyDown(event: React.KeyboardEvent<HTMLInputElement>) {
      if (event.key === keyCodes.ENTER) {
        handleInputValueChange();
      }
      if (event.key === keyCodes.ESCAPE) {
        setIsEditing(false);
        setInputValue(value);
      }
    }

    function handleChange(event: React.ChangeEvent<HTMLInputElement>) {
      setInputValue(event.target.value);
    }

    const clickableProps = useClickableProps(
      {
        onClick: toggleEditMode,
        id,
        role: "button"
      },
      mergedRef
    );

    function focus() {
      if (inputRef.current) {
        inputRef.current?.focus();
      }
    }

    function select() {
      if (inputRef.current) {
        inputRef.current.select();
      }
    }

    useEffect(() => {
      if (isEditing && focusOnMount) {
        focus();
      }
    }, [focusOnMount, isEditing]);

    useLayoutEffect(() => {
      if (isEditing && selectOnMount) {
        select();
      }
    }, [selectOnMount, isEditing]);

    const inputClassNames = useMemo(() => {
      return cx(styles.input, getStyle(styles, camelCase(type + "-" + weight)));
    }, [type, weight]);

    return (
      <div
        ref={mergedRef}
        id={id}
        aria-label={ariaLabel}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.EDITABLE_HEADING, id)}
        className={cx(styles.editableHeading, className)}
      >
        {isEditing && !readOnly ? (
          <>
            <HiddenInputPlaceholder className={inputClassNames} value={inputValue} onChange={setInputWidth} />
            <input
              ref={inputRef}
              className={inputClassNames}
              value={inputValue}
              onChange={handleChange}
              onKeyDown={handleKeyDown}
              onBlur={handleBlur}
              aria-label={ariaLabel}
              style={{ width: inputWidth }}
              role="input"
            />
          </>
        ) : (
          <Heading
            {...clickableProps}
            className={cx(styles.heading, { [styles.disabled]: readOnly })}
            type={type}
            weight={weight}
          >
            {inputValue}
          </Heading>
        )}
      </div>
    );
  }
);

export default withStaticProps(EditableHeading, {
  types: HeadingType,
  weights: HeadingWeight
});
