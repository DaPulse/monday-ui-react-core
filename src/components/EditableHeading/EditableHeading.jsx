import React, { useRef, useState, useCallback, useEffect, useMemo } from "react";
import PropTypes from "prop-types";
import cx from "classnames";
import { useButton } from "@react-aria/button";
import Heading from "../Heading/Heading";
import EditableInput, { TEXTAREA_TYPE } from "../EditableInput/EditableInput";
import { TYPES } from "../Heading/HeadingConstants";
import { SIZES } from "../../constants/sizes";
import usePrevious from "../../hooks/usePrevious";
import "./EditableHeading.scss";

const EditableHeading = props => {
  const {
    id,
    className,
    value,
    editing,
    disabled,
    onFinishEditing,
    onCancelEditing,
    errorClassTimeout,
    style,
    onStartEditing,
    contentRenderer
  } = props;

  // State
  const [isEditing, setIsEditing] = useState(editing && !disabled);
  const [isError, setIsError] = useState(false);
  const [valueState, setValueState] = useState(value || "");
  const prevValue = usePrevious(value);

  // Refs
  const ref = useRef(null);

  // Callbacks
  const onClick = useCallback(() => {
    if (disabled || isEditing) return;
    setIsEditing(true);
    onStartEditing && onStartEditing();
  }, [isEditing, disabled, setIsEditing, onStartEditing]);

  const onFinishEditingCallback = useCallback(
    (newValue, event) => {
      setIsEditing(false);
      setValueState(newValue);
      onFinishEditing && onFinishEditing(newValue, event);
    },
    [onFinishEditing, setIsEditing, setValueState]
  );

  const onCancelEditingCallback = useCallback(
    event => {
      setIsEditing(false);
      onCancelEditing && onCancelEditing(event);
    },
    [onCancelEditing, setIsEditing]
  );

  const clearErrorState = useCallback(() => {
    setIsError(false);
  }, [setIsError]);

  const onInputErrorCallback = useCallback(() => {
    setIsError(true);
  }, [setIsError]);

  const onInputSuccessCallback = useCallback(() => {
    clearErrorState();
  }, [clearErrorState]);

  const { buttonProps } = useButton({
    onPress: onClick,
    elementType: "div"
  });

  // Effects
  useEffect(() => {
    // Update value if changed from props
    if (!editing && value !== prevValue && value !== valueState) {
      setValueState(value);
    }
  }, [editing, value, prevValue, valueState, setValueState]);

  useEffect(() => {
    let timer;
    if (isError) {
      timer = setTimeout(clearErrorState, errorClassTimeout);
    }
    return () => timer && clearTimeout(timer);
  }, [isError, setIsError, clearErrorState, errorClassTimeout]);

  // Render
  const getContentProps = () => {
    const suggestEditOnHover = props.suggestEditOnHover && !disabled;
    const valueOrPlaceholder = valueState || props.placeholder || "";
    return {
      value: props.displayPlaceholderInTextMode ? valueOrPlaceholder : valueState,
      type: props.type,
      suggestEditOnHover,
      tooltipPosition: props.tooltipPosition,
      ellipsisMaxLines: props.ellipsisMaxLines,
      nonEllipsisTooltip: props.tooltip,
      size: props.size
    };
  };
  const renderContentComponent = () => {
    const contentProps = getContentProps();
    if (contentRenderer) {
      return contentRenderer(contentProps);
    }
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <Heading {...contentProps} />;
  };

  const getInputProps = () => {
    const textAreaType = props.ellipsisMaxLines > 1 ? TEXTAREA_TYPE : undefined;
    const inputType = props.inputType || textAreaType;
    return {
      value: valueState,
      className: `editable-heading-input element-type-${props.type} size-${props.size}`,
      isValidValue: props.isValidValue,
      onChange: props.onChange,
      onKeyDown: props.onKeyDown,
      onClick: props.onClick,
      onTabHandler: props.onTabHandler,
      onArrowKeyDown: props.onArrowKeyDown,
      autoComplete: props.autoComplete,
      maxLength: props.maxLength,
      placeholder: props.placeholder,
      shouldFocusOnMount: props.shouldFocusOnMount,
      selectOnMount: props.selectOnMount,
      inputType,
      ignoreBlurClass: props.ignoreBlurClass,
      autoSize: props.autoSize,
      textareaSubmitOnEnter: props.textareaSubmitOnEnter,
      onFinishEditing: onFinishEditingCallback,
      onCancelEditing: onCancelEditingCallback,
      onError: onInputErrorCallback,
      onSuccess: onInputSuccessCallback,
      ariaLabel: props.inputAriaLabel
    };
  };

  const renderInputComponent = () => {
    const inputProps = getInputProps();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <EditableInput {...inputProps} />;
  };

  const clickProps = useMemo(() => {
    return isEditing ? { onClick, role: "button", tabIndex: "0" } : buttonProps;
  }, [isEditing, buttonProps, onClick]);

  return (
    <div
      ref={ref}
      style={style}
      className={cx("editable-heading--wrapper", className)}
      {...clickProps}
      aria-label={`${value} ${props.tooltip || ""}`}
      id={id}
    >
      {isEditing ? renderInputComponent() : renderContentComponent()}
    </div>
  );
};

EditableHeading.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string,
  type: PropTypes.oneOf(Object.keys(TYPES)),
  errorClass: PropTypes.string,
  errorClassTimeout: PropTypes.number,
  displayPlaceholderInTextMode: PropTypes.bool,
  suggestEditOnHover: PropTypes.bool,
  autoSize: PropTypes.bool,
  size: PropTypes.oneOf(Object.values(SIZES)),
  inputAriaLabel: PropTypes.string
};
EditableHeading.defaultProps = {
  className: "",
  id: "",
  type: TYPES.H1,
  errorClass: "error",
  errorClassTimeout: 2000,
  displayPlaceholderInTextMode: true,
  suggestEditOnHover: true,
  autoSize: true,
  size: SIZES.LARGE,
  inputAriaLabel: undefined
};

EditableHeading.types = TYPES;
EditableHeading.sizes = SIZES;

export default EditableHeading;
