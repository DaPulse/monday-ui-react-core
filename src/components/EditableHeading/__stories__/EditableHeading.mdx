import EditableHeading from "../EditableHeading";
import { ArgsTable, Canvas, Meta, Story } from "@storybook/blocks";
import { createStoryMetaSettingsDecorator } from "../../../storybook";
import { createComponentTemplate } from "vibe-storybook-components";
import {
  EDITABLE_TEXT,
  HEADING,
  TEXT
} from "../../../storybook/components/related-components/component-description-map";
import * as EditableHeadingStories from "./EditableHeading.stories";

export const metaSettings = createStoryMetaSettingsDecorator({
  component: EditableHeading,
  enumPropNamesArray: ["weight", "type"],
  actionPropsArray: ["onChange"]
});

<Meta of={EditableHeadingStories} />

export const editableHeadingTemplate = createComponentTemplate(EditableHeading);

# EditableHeading

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)
- [Related components](#related-components)
- [Feedback](#feedback)

## Overview

Editable Heading component is based on the <Link href="/?path=/docs/text-heading--overview" withoutSpacing>Heading</Link> component, allowing users to dynamically edit its content.

<Canvas>
  <Story of={EditableHeadingStories.Overview} />
</Canvas>

## Props

<ArgsTable story="Overview" />

## Usage

<UsageGuidelines
  guidelines={[
    "Use when you want to allow the user to edit the heading text",
    "This component grows relative to its content, and its maximum width is 100% of its container "
  ]}
/>

## Variants

### Heading types

<Canvas>
  <Story of={EditableHeadingStories.Types} />
</Canvas>

## Related components

<RelatedComponents componentsNames={[EDITABLE_TEXT, TEXT, HEADING]} />
