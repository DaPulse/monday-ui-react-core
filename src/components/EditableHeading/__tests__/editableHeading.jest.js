import React from "react";
import renderer from "react-test-renderer";
import { render, cleanup } from "@testing-library/react";
import EditableHeading from "../EditableHeading";
import "@testing-library/jest-dom";

describe("EditableHeading", () => {
  afterEach(() => {
    cleanup();
  });

  it("renders correctly with empty props", () => {
    const tree = renderer.create(<EditableHeading type={EditableHeading.types.h1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should not render an button when editing = false", () => {
    const { queryByRole } = render(<EditableHeading type={EditableHeading.types.h1} editing={false} />);

    const button = queryByRole('button');

    expect(button).toBeNull();
  });

  it("should render a button when editing = true", () => {
    const { getByRole } = render(<EditableHeading type={EditableHeading.types.h1} editing={true} />);

    const button = getByRole('button');

    expect(button).toBeInTheDocument();
  });

  it("should render a button when editing prop changes from false to true", () => {
    const { getByRole, rerender } = render(<EditableHeading type={EditableHeading.types.h1} />);

    rerender(<EditableHeading type={EditableHeading.types.h1} editing={true} />);
    const button = getByRole('button');

    expect(button).toBeInTheDocument();
  });
});
