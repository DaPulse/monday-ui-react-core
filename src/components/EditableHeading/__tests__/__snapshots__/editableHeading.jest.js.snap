// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with empty props 1`] = `
<div
  aria-label="undefined "
  className="editable-heading--wrapper"
  id=""
  onClick={[Function]}
  onDragStart={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <h1
    aria-label=""
    className="heading-component element-type-h1 size-large heading-element-ellipsis suggest-edit-on-hover"
    id=""
  >
    
  </h1>
</div>
`;
